src/protocol/sort/secureapplyinv.rs

```diff
@@ -0,0 +1,124 @@
+#[derive(Debug)]
+#[embed_doc_image("secureapplyinv", "images/sort/secureapplyinv.png")]
+pub struct SecureApplyInv {}
+
+impl SecureApplyInv {
+    /// This algorithm applies a permutation to the `input` vector. The permutation is secret-shared,
+    /// and none of the helpers should learn it through this protocol.
+    /// To keep the permutation secret, it (and the inputs) are first randomly securely shuffled.
+    /// After this shuffle, the permutation can be revealed.
+    /// An adversary can only obtain a shuffled permutation, which is just a random permutation.
+    /// Steps
+    /// ![Secure Apply Inv steps][secureapplyinv]
+    /// 1. Generate random permutations using prss
+    /// 2. Secret shared permutation is shuffled with random permutations
+    /// 3. Secret shared value is shuffled using the same random permutations
+    /// 4. The permutation is revealed
+    /// 5. All helpers call `apply` to apply the permutation locally.
+    #[allow(dead_code)]
+    pub async fn execute<F: Field>(
+        ctx: &ProtocolContext<'_, F>,
+        input: &'_ mut Vec<Replicated<F>>,
+        sort_permutation: &'_ mut Vec<Replicated<F>>,
+    ) -> Result<(), BoxError> {
+        let mut random_permutations =
+            get_two_of_three_random_permutations(input.len(), &ctx.prss());
+        let mut random_permutations_copy = random_permutations.clone();
+
+        let mut shuffle_inputs = Shuffle::new(input);
+        let mut shuffle_permutation = Shuffle::new(sort_permutation);
+
+        let (_, _) = try_join!(
+            shuffle_inputs.execute(ctx.narrow(&ShuffleInputs), &mut random_permutations,),
+            shuffle_permutation.execute(
+                ctx.narrow(&ShufflePermutation),
+                &mut random_permutations_copy
+            ),
+        )?;
+        let mut permutation =
+            reveal_a_permutation(ctx.narrow(&RevealPermutation), sort_permutation).await?;
+
+        apply(&mut permutation, input);
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use permutation::Permutation;
+    use rand::seq::SliceRandom;
+    use tokio::try_join;
+
+    use crate::{
+        protocol::{sort::apply::apply_inv, QueryId},
+        test_fixture::{generate_shares, make_contexts, make_world, validate_result_from_shares},
+    };
+
+    use super::SecureApplyInv;
+
+    #[tokio::test]
+    pub async fn secureapplyinv() {
+        const BATCHSIZE: usize = 25;
+        for _ in 0..10 {
+            let mut rng = rand::thread_rng();
+            let input: Vec<u128> = (0..(BATCHSIZE as u128)).collect();
+
+            let mut permutation: Vec<usize> = (0..BATCHSIZE).collect();
+            permutation.shuffle(&mut rng);
+
+            let mut expected_result = input.clone();
+            let mut cloned_perm = Permutation::oneline(permutation.clone()).inverse();
+
+            apply_inv(&mut cloned_perm, &mut expected_result);
```

Comment:
I find this extremely difficult to follow.

It's really strange that `expected_result` is mutable. Maybe some comments would help? All of the cloning definitely leaves me saying "huh? what just happenned?" And this apply_inv of the inverse of something... all of these "inverses" confuse me a lot.

Reply:
Since SecureApplyInv changes the inputs sent to it, ended up cloning them to validate the results.
The actual paper expects us to apply an inverse on the inverted Permutation (i.e. apply_inv(perm.inverse(), input)).
Since this is same as apply(perm, input), we are doing this instead now both in the code and in the test.
Added this comment to reflect correctly